docs/src/neptune-core/overview.md

```diff
@@ -67,25 +66,38 @@ Different tasks are scheduled in the main loop every N seconds. These currently
 
-Blocks are stored on disk and their position on disk is stored in the `block_index` database. Blocks are read from and written to disk using `mmap`.
+We use `leveldb` for our database layer with custom wrappers that make it more async-friendly, type safe, and emulate multi-table transactions.
+
+`neptune_core::database::NeptuneLevelDb` provides async wrappers for leveldb APIs to avoid blocking async tasks.
+
+`leveldb` is a simple key/value store, meaning it only allows manipulating individual strings. It does however provide a batch update facility. `neptune_core::database::storage::storage_schema::DbSchema` leverages these batch updates to provide vector and singleton types that can be manipulated in rust code and then atomically written to `leveldb` as a single batch update (aka transaction).
+
+Blocks are stored on disk and their position on disk is stored in the `block_index` database. Blocks are read from and written to disk using `mmap`.  We wrap all file-system calls with tokio's `spawn_blocking()` so they will not block other async tasks.
 
 ## Challenges
-- Deadlocks. Solution: always acquire locks in the same order. Note though that locks from `std::sync` may not be held over an `await`. The linter should tell you if you do this. When a function requires more than one lock, **the only correct ordering in which to acquire these locks is the order in which the fields are defined in `GlobalState`. Any deviation from this is a bug.**
+
+- Deadlocks. We only have a single RwLock over the GlobalState. This is encapsulated in struct `GlobalStateLock`. This makes deadlocks pretty easy to avoid, following some simple rules:
+
+  1. avoid deadlocking yourself. If a function has read-acquired the global lock then it must be released before write-acquiring. Likewise never attempt to write-acquire the lock twice.
+
+  2. avoid deadlocking others. Always be certain that the global lock will be released in timely fashion. In other words if you have some kind of long running task with an event loop that needs to acquire the global lock, ensure that it gets acquired+released inside the loop rather than outside.
+
 - Atomic writing to databases: The archival mutator set is spread across multiple databases due to how the underlying data structures are defined. If one of the databases are updated but the other is not, this will leave the archival mutator set in an invalid state. We could fix this by allowing an archival mutator set to be stored in only one database. We should also add logic to rebuild the archival mutator set state from the `block_index` database and the blocks stored on disk since it can be derived from the blocks. This functionality could be contained in a separate binary, just like we have a binary for the CLI interface in the form of the RPC client.
```

Comment:
We manage the entire mutator set through *one* database now, so the 1st sentence is now incorrect. However, this warning/problem is true for the `ArchivalState`, so maybe we can just replace "archival mutator set" with "archival state" here, and it still holds?

Please note, I am not advocating or suggesting any change to the code. I'm happy with how the state looks, this is just about the documentation.

Reply:
ok, I changed it.

Reply:
Let me add a few considerations:

Atomic writing to databases: neptune-core presently writes to the following databases: wallet_db, block_index_db, archival_mutator_set, peer_state. If one of the databases are updated but the other is not, this can leave data in an invalid state. We could fix this by storing all state in a single transactional database *but that might make the code base less modular.*

note: We should also add logic to rebuild the archival state from the block_index_db and the blocks stored on disk since it can be derived from the blocks. This functionality could be contained in a separate binary, just like we have a binary for the CLI interface in the form of the RPC client, *or this could be a check that was performed at startup.*

Reply:
Ok, done. I shortened the last sentence of 2nd para to:

This functionality could be contained in a separate binary or a check could be performed at startup.